utils/vlog/vlog_to_pbtype.py

```diff
@@ -55,294 +58,531 @@
+    all_have = []
+    for attr in set(sum((list(s.keys()) for s in srcs), [])):
+        if len(srcs) != sum(bool(attr in s) for s in srcs):
+            continue
+        all_have.append(attr)
 
-def strip_name(name):
-    if '\\' in name:
-        ts = name.find('\\')
-        tf = name.rfind('\\')
-        return name[ts + 1:tf]
-    return name
+    for attr in all_have:
+        avalue = srcs[0][attr]
+        avalues = [s[attr] for s in srcs[1:]]
+
+        for other_avalue in avalues:
+            if avalue == other_avalue:
+                continue
+            raise ValueError('{} values: {}'.format(attr, [avalue] + avalues))
+
+        if attr in dst:
+            assert avalue == dst[attr]
+            raise ValueError(
+                '{} on net has value {} but pins have {}'.format(
+                    attr, dst[attr], avalue
+                )
+            )
+        dst[attr] = avalue
+
+
+def net_and_pin_attrs(yj, mod, driver: CellPin, sink: CellPin, netid: int):
+    def filter_src(x):
+        d = {}
+        for k, v in x.items():
+            if k == 'src':
+                continue
+            d[k] = v
+        return d
+
+    potential_attrs = []
+
+    driver_cell, driver_pin = driver
+    driver_type = mod.cell_type(driver_cell)
+    if driver_type != None:
+        dmod = yj.module(driver_type)
+        potential_attrs.append(filter_src(dmod.port_attrs(driver_pin)))
+
+    sink_cell, sink_pin = sink
+    sink_attrs = {}
+    sink_type = mod.cell_type(sink_cell)
+    if sink_type != None:
+        smod = yj.module(sink_type)
+        potential_attrs.append(filter_src(smod.port_attrs(sink_pin)))
+
+    net_attrs = filter_src(mod.net_attrs(mod.net_name(netid)))
+    copy_attrs(net_attrs, potential_attrs)
+    return net_attrs
+
+
+def make_direct_conn(
+        ic_xml: ET.Element, driver: CellPin, sink: CellPin, path_attr: dict
+) -> ET.Element:
+    dir_xml = ET.SubElement(ic_xml, 'direct')
+    create_port(dir_xml, driver, "input")
+    create_port(dir_xml, sink, "output")
+
+    pack_name = path_attr.get('pack', False)
+    if pack_name:
+        pp_xml = ET.SubElement(
+            dir_xml, 'pack_pattern', {
+                'name': pack_name,
+                'type': 'pack'
+            }
+        )
+        create_port(pp_xml, driver, "input")
+        create_port(pp_xml, sink, "output")
+
+    carry_name = path_attr.get('carry', None)
+    if carry_name:
+        pp_xml = ET.SubElement(
+            dir_xml, 'pack_pattern', {
+                'name': carry_name,
+                'type': 'carry'
+            }
+        )
+        create_port(pp_xml, driver, "input")
+        create_port(pp_xml, sink, "output")
+
+    return dir_xml
+
+
+def make_mux_conn(
+        ic_xml: ET.Element, mux_name: str, mux_inputs: Dict[CellPin, CellPin],
+        mux_outputs: Dict[CellPin, List[CellPin]]
+) -> ET.Element:
+
+    ## <mux input="use_mux.a use_mux.b use_mux.cin" name="AMUX" output="use_mux.cout">
```

Comment:
What is going on with this giant comment?  If it is doctest like, just make a doctest?

Maybe better suited as a block quote?

Reply:
This was my first approach at implementing this functionality that didn't quite work. Will remove shortly.

Reply:
@kgugala

Reply:
this comment has been removed